File: dukechess/js-backend/movement/Summon.js
const MovementImpl = require("./MovementImpl");
const DestUtils = require("./DestUtils");

class Summon extends MovementImpl {
	validateMovement(field, pos, d, p) {
		var ret = [];
		
		var offsets = DestUtils.dest2Offset(d, p);
		
		if (0 != offsets[0] & 0 != offsets[1] & Math.abs(offsets[0]) != Math.abs(offsets[1])) return ret;
		
		if (!(1 == Math.abs(offsets[0]) || 1 == Math.abs(offsets[1]))) return ret;
		
		if (!this.isInField(field, pos, offsets) || this.hasAnyChess(field, pos, offsets)) return ret;
		
		ret = ret.concat(this.offset2Dest(field, pos, offsets));
		
		return ret;
	}

	validateControlArea(field, pos, d, p) {
		return [];
	}
}

module.exports = Summon;

File: deprecated/Java/com.dexaint.dukechess/src/main/java/com/dexaint/dukechess/movement/Summon.java
package com.dexaint.dukechess.movement;

import org.apache.commons.lang3.ArrayUtils;

import com.dexaint.dukechess.flow.Field;
import com.dexaint.dukechess.flow.Player;

public class Summon extends MovementImpl {
	public int[] validateMovement(Field field, int pos, Destination d, Player p) {
		int[] ret = new int[]{};
		
		int[] offsets = d.dest2Offset(p);
		
		if (0 != offsets[0] & 0 != offsets[1] & Math.abs(offsets[0]) != Math.abs(offsets[1])) return ret;
		
		if (!(1 == Math.abs(offsets[0]) || 1 == Math.abs(offsets[1]))) return ret;
		
		if (!isInField(field, pos, offsets) || hasAnyChess(field, pos, offsets)) return ret;
		
		ret = ArrayUtils.add(ret, offset2Dest(field, pos, d.dest2Offset(p)));
		
		return ret;
	}
}
